src/aspire/abinitio/commonline_sdp.py

```diff
@@ -1,30 +1,249 @@
+        v2 = v[self.n_img : 2 * self.n_img].T
+
+        # Use a least-squares method to get A.T*A and a Cholesky decomposition to find A.
+        A = self._ATA_solver(v1, v2)
+
+        # Recover the rotations. The first two columns of all rotation
+        # matrices are given by unmixing V1 and V2 using A. The third
+        # column is the cross product of the first two.
+        r1 = np.dot(A.T, v1)
+        r2 = np.dot(A.T, v2)
+        r3 = np.cross(r1, r2, axis=0)
+
+        rotations = np.empty((self.n_img, 3, 3), dtype=self.dtype)
+        rotations[:, :, 0] = r1.T
+        rotations[:, :, 1] = r2.T
+        rotations[:, :, 2] = r3.T
+        # Make sure that we got rotations by enforcing R to be
+        # a rotation (in case the error is large)
+        u, _, v = np.linalg.svd(rotations)
+        np.einsum("ijk, ikl -> ijl", u, v, out=rotations)
+
+        return rotations
+
+    @staticmethod
+    def _ATA_solver(v1, v2):
         """
-        Output the 3D orientations
+        Uses a least squares method to solve for the linear transformation A
+        such that A*v1=R1 and A*v2=R2 are the columns of the rotations matrices.
+
+        :param v1: 3 x n_img array corresponding to linear combinations of the first
+            columns of all rotation matrices.
+        :param v2: 3 x n_img array corresponding to linear combinations of the second
+            columns of all rotation matrices.
+
+        :return: 3x3 linear transformation mapping v1, v2 to first two columns of rotations.
         """
-        pass
+        # We look for a linear transformation (3 x 3 matrix) A such that
+        # A*v1'=R1 and A*v2=R2 are the columns of the rotations matrices.
+        # Therefore:
+        # v1 * A'*A v1' = 1
+        # v2 * A'*A v2' = 1
+        # v1 * A'*A v2' = 0
+        # These are 3*K linear equations for 9 matrix entries of A'*A
+        # Actually, there are only 6 unknown variables, because A'*A is symmetric.
+        # So we will truncate from 9 variables to 6 variables corresponding
+        # to the upper half of the matrix A'*A
+        n_img = v1.shape[-1]
+        truncated_equations = np.zeros((3 * n_img, 9), dtype=v1.dtype)
+        k = 0
+        for i in range(3):
+            for j in range(3):
+                truncated_equations[0::3, k] = v1[i] * v1[j]
+                truncated_equations[1::3, k] = v2[i] * v2[j]
+                truncated_equations[2::3, k] = v1[i] * v2[j]
+                k += 1
+
+        # b = [1 1 0 1 1 0 ...]' is the right hand side vector
+        b = np.ones(3 * n_img)
+        b[2::3] = 0
+
+        # Find the least squares approximation of A'*A in vector form
+        ATA_vec = np.linalg.lstsq(truncated_equations, b, rcond=None)[0]
+
+        # Construct the matrix A'*A from the vectorized matrix.
+        # Note, this is only the lower triangle of A'*A.
+        ATA = ATA_vec.reshape(3, 3)
+
+        # The Cholesky decomposition of A'*A gives A (lower triangle).
```

Comment:
Might be worthwhile to note that `np.linalg.cholesky` only needs the lower triangle of `ATA`.

Reply:
Sure. Done!